### PR TITLE
[dcl.fct.def.coroutine] Initializer of a non-block variable is also a resumer CWG2613

### LINKED ISSUE #5295: The resumer of a coroutine may not be a function
````cpp
Task task(){
   std::cout<<"in task\n";
   int r = co_await Line();
   std::cout<<"resumed\n";
   co_return r;
}
auto r = task();
auto c = (r.coro_.resume(),0);  // #1
auto f = [](){
    std::cout<<"invoked\n";
    return 0;
}();
int main(){}
````
`#1` is a declaration where its initializer expression will invoke the resumption member function of the coroutine. We just state that:
> A suspended coroutine can be resumed to continue execution by invoking a resumption member function ([coroutine.handle.resumption]) of a coroutine handle ([coroutine.handle]) that refers to the coroutine. **The function that invoked a resumption member function is called the resumer**.  

and we have specified that
> A coroutine returns to its caller or **resumer** ([dcl.fct.def.coroutine]) by the co_­return statement or when suspended ([expr.await]).   

Obviously, in this case, the call of the resumption member function at `#1` causes the coroutine to be resumed, and the *co_return* statement can cause the coroutine to return to its resumer, which means the **lambda** will be subsequently invoked. The complete example is [here](https://godbolt.org/z/9nqP5a89M). The implementation admits that the context at `#1` is also a so-called resumer.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #5295

### CODE REVIEW

File: source/declarations.tex [hunk 1]
```diff
@@ -6554,8 +6554,8 @@
 a resumption member function\iref{coroutine.handle.resumption}
 of a coroutine handle\iref{coroutine.handle}
 that refers to the coroutine.
-The function that invoked a resumption member function is
-called the \defnx{resumer}{coroutine!resumer}.
+The function or the initializer of a non-block variable
+that invoked a resumption member function is called the \defnx{resumer}{coroutine!resumer}.
```
Comment (xmh0511) [inline source/declarations.tex:6558]: maybe "that invokes a resumption member function" is better.

**Comment by xmh0511** [comment]
There are also two special cases:
````cpp
/* case 1 */
int fun1(int a = (r.resume(), 0)){
    return a;
}
int v1 = fun1();

/* case 2*/
int fun2(int a){
   return a;
}
int v2 = fun2((r.resume(),0));
````
First, any function parameter belongs to function parameter scope, thus the variables introduced by **[parameter-declaration](https://eel.is/c++draft/dcl.fct#nt:parameter-declaration)** are non-block variables. This proposal intends to use **initializer** to clarify the issue. However, the clear point in the two cases is that the initializer, in either case, is `= E` as per the grammar *[init-declarator](https://eel.is/c++draft/dcl.decl#nt:init-declarator)* (declarator **initializer**). The subtle point is whether we consider the **argument** of the call, in either case, as the resumer? Although, the argument is not a grammatical initializer(BTW, the issue regarding the initializer is recorded in https://github.com/cplusplus/draft/issues/4582 and https://github.com/cplusplus/draft/issues/4584)

**Comment by frederick-vs-ja** [comment]
I don't find anything ambiguous. In these two case, the `resume` calls appear in expressions that initialize other variables, but they are also executed within the initializers of non-block ("global") variables, and outside of functions.

**Comment by xmh0511** [comment]
> I don't find anything ambiguous. In these two case, the `resume` calls appear in expressions that initialize other variables, but they are also executed within the initializers of non-block ("global") variables, and outside of functions.

According to the current proposal, the initializers `= fun1()` and `= fun2((r.resume(),0))` in the two declarations of the global variables are resumers, there are no hazy here. I just wonder Is it possible that the argument that initializes the parameter in the context that is not a block can be a resumer?

**Comment by frederick-vs-ja** [comment]
>I just wonder Is it possible that the argument that initializes the parameter in the context that is not a block can be a resumer?

I think this is a question of classification. When the `resume` function is called within another function, the existing wording doesn't treat any smaller structure (e.g. a expression statement) in that function as a resumer.
So IMO an expression in the initializer of a parameter should not be treated as a resumer. The resumer in this case should be one of these two enclosing contexts.

**Comment by xmh0511** [comment]
> So IMO an expression in the initializer of a parameter should not be treated as a resumer.

If we do not intend to treat the initializer of the parameter as a resumer, the current proposal should be modified as
> The function or the initializer of a variable that invokes a resumption member function, where the variable does not belong to the block or parameter scope, is called the resumer.  

the variable that belongs to function parameter scope is also a non-block variable as per [basic.scope.block] p1
> A variable that belongs to a block scope is a [block variable](https://eel.is/c++draft/basic.scope.block#def:block_variable)[.](https://eel.is/c++draft/basic.scope.block#1.sentence-3)

This also works if we clarify the definition of the wording "initializer" after fixing issues  https://github.com/cplusplus/draft/issues/4582 and https://github.com/cplusplus/draft/issues/4584

**Comment by frederick-vs-ja** [comment]
I think the clarification of "non-block variable" should be handled in another issue/PR. [[basic.start.dynamic]](https://eel.is/c++draft/basic.start.dynamic) frequently use the item "non-block variable", this PR just reuse the meaning that section. I'll added a cross reference for this use.

**Comment by xmh0511** [comment]
[basic.start.dynamic] explicitly states: a non-block variable with **static storage duration**, however, we do not add that condition to non-block variable in this revision.

**Comment by frederick-vs-ja** [comment]
> [basic.start.dynamic] explicitly states: a non-block variable with **static storage duration**. However, we did not add that condition for the non-block variable in this PR.

Thanks. It was my oversight.
The current working draft almost always combines "non-block variable" with "with static (or thread) storage duration", and never uses the item to denote a function parameter. I should follow this (perhaps slightly redundant) pattern.

**Comment by tkoeppe** [comment]
@GorNishanov, @lewissbaker could you kindly take a look?

**Comment by GorNishanov** [comment]
I am wondering if there is a more general wording than trying to enumerate all contexts that may resume the coroutine.

**Comment by jensmaurer** [comment]
I'm wondering whether this should go through CWG. Maybe we can just say "context" instead of "function or x or y or z" instead.

**Comment by tkoeppe** [comment]
@jensmaurer I would welcome that.

**Comment by lewissbaker** [comment]
> I am wondering if there is a more general wording than trying to enumerate all contexts that may resume the coroutine.

I think we are just trying to say that when a coroutine suspends we return from the call to `coroutine_handle::resume()` that resumed the coroutine.

So can we say something along the following lines?:
> The coroutine _resumer_ is the call-site of the call to `coroutine_handle::resume()` that resumed the coroutine.

**Comment by jensmaurer** [comment]
I made CWG2613 for this. Better get larger buy-in for whatever term we end up with.

**Comment by jensmaurer** [comment]
I'm closing this pull-request; it's unlikely the core issue resolution will use exactly the phrasing proposed here.

**Pull request closed without merging**